File: frontend/src/api/backup.ts
import api from './index'

export interface BackupConfig {
  id: number
  family_id: number
  backup_type: string
  schedule: string
  is_enabled: boolean
  target: string
  target_config: Record<string, unknown> | null
  retention_days: number
  max_backups: number
}

export interface BackupLog {
  id: number
  backup_type: string
  backup_target: string
  file_path: string | null
  file_size: number | null
  file_format: string | null
  table_counts: Record<string, number | string> | null
  status: string
  error_message: string | null
  duration_ms: number | null
  created_at: string | null
}

export interface BackupConfigCreate {
  backup_type: string
  schedule: string
  target: string
  target_config?: Record<string, unknown>
  retention_days?: number
  max_backups?: number
}

export function getBackupConfigs() {
  return api.get<BackupConfig[]>('/backup/configs')
}

export function createBackupConfig(data: BackupConfigCreate) {
  return api.post<BackupConfig>('/backup/configs', data)
}

export function triggerBackup() {
  return api.post<BackupLog>('/backup/trigger')
}

export function getBackupLogs() {
  return api.get<BackupLog[]>('/backup/logs')
}

export function deleteBackupLog(logId: number) {
  return api.delete(`/backup/logs/${logId}`)
}

File: backend/app/models/backup.py
from sqlalchemy import BigInteger, Boolean, DateTime, Integer, String
from sqlalchemy.dialects.postgresql import JSONB
from sqlalchemy.orm import Mapped, mapped_column
from sqlalchemy.sql import func

from app.models.base import Base


class BackupConfig(Base):
    __tablename__ = "backup_configs"

    id: Mapped[int] = mapped_column(BigInteger, primary_key=True, autoincrement=True)
    family_id: Mapped[int] = mapped_column(BigInteger, nullable=False)
    backup_type: Mapped[str] = mapped_column(String(30), nullable=False)
    schedule: Mapped[str] = mapped_column(String(50), nullable=False)
    is_enabled: Mapped[bool] = mapped_column(Boolean, default=True)
    target: Mapped[str] = mapped_column(String(30), nullable=False)
    target_config: Mapped[dict | None] = mapped_column(JSONB)
    retention_days: Mapped[int] = mapped_column(Integer, default=30)
    max_backups: Mapped[int] = mapped_column(Integer, default=10)
    created_at: Mapped[str] = mapped_column(DateTime(timezone=True), server_default=func.now())
    updated_at: Mapped[str] = mapped_column(DateTime(timezone=True), server_default=func.now(), onupdate=func.now())


class BackupLog(Base):
    __tablename__ = "backup_logs"

    id: Mapped[int] = mapped_column(BigInteger, primary_key=True, autoincrement=True)
    backup_type: Mapped[str] = mapped_column(String(30), nullable=False)
    backup_target: Mapped[str] = mapped_column(String(30), nullable=False)
    file_path: Mapped[str | None] = mapped_column(String(500))
    file_size: Mapped[int | None] = mapped_column(BigInteger)
    file_format: Mapped[str | None] = mapped_column(String(20))
    table_counts: Mapped[dict | None] = mapped_column(JSONB)
    status: Mapped[str] = mapped_column(String(20), default="success")
    error_message: Mapped[str | None] = mapped_column(String(500))
    duration_ms: Mapped[int | None] = mapped_column(Integer)
    created_at: Mapped[str] = mapped_column(DateTime(timezone=True), server_default=func.now())
